Use more ICU surrogate code point macros

annevk opened the pull request:
#### b58c8bed4445a0192e15f9bdc67bd0f0c45d4ea2
<pre>
Use more ICU surrogate code point macros
<a href="https://bugs.webkit.org/show_bug.cgi?id=254984">https://bugs.webkit.org/show_bug.cgi?id=254984</a>
rdar://107602109

Reviewed by Darin Adler.

Make WebKit rely more on centrally defined checks and constants for code points, in particular for surrogates.

* Source/JavaScriptCore/runtime/JSImmutableButterfly.cpp:
(JSC::JSImmutableButterfly::createFromString):
* Source/JavaScriptCore/runtime/RegExpObjectInlines.h:
(JSC::advanceStringUnicode):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::utf8ForCharactersIntoBuffer):
* Source/WebCore/PAL/pal/text/TextCodec.cpp:
(PAL::TextCodec::getUnencodableReplacement):
* Source/WebCore/css/parser/CSSTokenizer.cpp:
(WebCore::CSSTokenizer::consumeEscape):
* Source/WebCore/dom/TextEncoderStreamEncoder.cpp:
(WebCore::TextEncoderStreamEncoder::encode):
(WebCore::TextEncoderStreamEncoder::flush):
* Source/WebCore/dom/TextEncoderStreamEncoder.h:

Canonical link: <a href="https://commits.webkit.org/262581@main">https://commits.webkit.org/262581@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e96a07399951c1d5e9567c6c7f2c6965aec5624e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1940 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1971 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2033 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2867 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2028 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1916 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2048 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2017 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1794 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1959 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1747 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1739 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2713 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1738 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1722 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1638 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/1629 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1784 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1769 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2887 "10 api tests failed or timed out") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/1859 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1796 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1599 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/2000 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1723 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1733 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/456 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/472 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1887 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/2045 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/571 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/dom/TextEncoderStreamEncoder.cpp:
```diff
@@ -44,22 +45,21 @@ RefPtr<Uint8Array> TextEncoderStreamEncoder::encode(const String& input)
-                auto codePoint = 0x10000 + ((highSurrogate - 0xD800) << 10) + (token - 0xDC00);
+        if (m_pendingLeadSurrogate) {
+            auto leadSurrogate = *std::exchange(m_pendingLeadSurrogate, std::nullopt);
+            if (U_IS_TRAIL(token)) {
```
I think this is a code unit and not a code point and so should be `U16_IS_TRAIL`. In practice either macro will work.

annevk replied:
Thanks, addressed all of these. I also addressed this in StringImpl.cpp where it applied as well.

darinadler commented on Source/WebCore/dom/TextEncoderStreamEncoder.cpp:
```diff
@@ -44,22 +45,21 @@ RefPtr<Uint8Array> TextEncoderStreamEncoder::encode(const String& input)
         }
-        if (token >= 0xD800 && token <= 0xDBFF) {
-            m_pendingHighSurrogate = token;
+        if (U_IS_LEAD(token)) {
```
Same thought here.

darinadler commented on Source/WebCore/dom/TextEncoderStreamEncoder.cpp:
```diff
@@ -44,22 +45,21 @@ RefPtr<Uint8Array> TextEncoderStreamEncoder::encode(const String& input)
         }
-        if (token >= 0xDC00 && token <= 0xDFFF) {
-            U8_APPEND_UNSAFE(bytes.data(), bytesWritten, 0XFFFD);
+        if (U_IS_TRAIL(token)) {
```
And here.

annevk committed:
Use more ICU surrogate code point macros
https://bugs.webkit.org/show_bug.cgi?id=254984
rdar://107602109

Reviewed by Darin Adler.

Make WebKit rely more on centrally defined checks and constants for code points, in particular for surrogates.

* Source/JavaScriptCore/runtime/JSImmutableButterfly.cpp:
(JSC::JSImmutableButterfly::createFromString):
* Source/JavaScriptCore/runtime/RegExpObjectInlines.h:
(JSC::advanceStringUnicode):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::utf8ForCharactersIntoBuffer):
* Source/WebCore/PAL/pal/text/TextCodec.cpp:
(PAL::TextCodec::getUnencodableReplacement):
* Source/WebCore/css/parser/CSSTokenizer.cpp:
(WebCore::CSSTokenizer::consumeEscape):
* Source/WebCore/dom/TextEncoderStreamEncoder.cpp:
(WebCore::TextEncoderStreamEncoder::encode):
(WebCore::TextEncoderStreamEncoder::flush):
* Source/WebCore/dom/TextEncoderStreamEncoder.h:

Canonical link: https://commits.webkit.org/262581@main
